[SQL] Arithmetic operations on types TIME and INTERVAL fail to compile

For a SQL program such as:
```
CREATE TABLE time_tbl(
c1 TIME);

CREATE MATERIALIZED VIEW v AS SELECT
c1 - INTERVAL '10' MINUTE AS f_c1
FROM time_tbl;
```

The compilation fails with an error:
**error[E0425]: cannot find function `minus_TimeN_ShortInterval` in this scope**

--- a/crates/sqllib/src/interval.rs
+++ b/crates/sqllib/src/interval.rs
@@ -3,16 +3,19 @@
 
 use crate::{
     operators::{eq, gt, gte, lt, lte, neq},
-    some_existing_operator, some_operator,
+    some_existing_operator, some_operator, some_polymorphic_function2, Decimal,
 };
-use dbsp::num_entries_scalar;
+use dbsp::{algebra::F64, num_entries_scalar};
 use feldera_types::serde_with_context::{
     DeserializeWithContext, SerializeWithContext, SqlSerdeConfig,
 };
 use num::PrimInt;
 use serde::{de::Error as _, ser::Error as _, Deserialize, Serialize};
 use size_of::SizeOf;
-use std::{fmt::Debug, ops::Mul};
+use std::{
+    fmt::Debug,
+    ops::{Div, Mul},
+};
 
 #[cfg(doc)]
 use crate::{Date, Time, Timestamp};
@@ -116,18 +119,73 @@ pub fn to_bound_ShortInterval_Time_u64(value: &ShortInterval) -> u64 {
     (value.milliseconds * 1_000_000) as u64
 }
 
-/// Multiply a `ShortInterval` by a numeric value, producing a
-/// `ShortInterval`.
-impl<T> Mul<T> for ShortInterval
-where
-    T: PrimInt,
-    i64: Mul<T, Output = i64>,
-{
+/// Multiply a `ShortInterval` by a numeric value, producing a `ShortInterval`.
+impl Mul<i64> for ShortInterval {
     type Output = Self;
 
-    fn mul(self, rhs: T) -> Self {
+    /// Multiply a short interval by an long integer
+    fn mul(self, rhs: i64) -> Self {
         Self {
             milliseconds: self.milliseconds * rhs,
+        }
+    }
+}
+
+/// Multiply a `ShortInterval` by a numeric value, producing a `ShortInterval`.
+impl Mul<F64> for ShortInterval {
+    type Output = Self;
+
+    fn mul(self, rhs: F64) -> Self {
+        Self {
+            milliseconds: (F64::from(self.milliseconds as f64) * rhs).into_inner() as i64,
+        }
+    }
+}
+
+/// Multiply a `ShortInterval` by a numeric value, producing a `ShortInterval`.
+impl Mul<Decimal> for ShortInterval {
+    type Output = Self;
+
+    fn mul(self, rhs: Decimal) -> Self {
+        Self {
+            milliseconds: (Decimal::from(self.milliseconds) * rhs)
+                .try_into()
+                .expect("overflow in short interval multiplication"),
+        }
+    }
+}
+
+/// Divide a `ShortInterval` by a numeric value, producing a `ShortInterval`.
+impl Div<i64> for ShortInterval {
+    type Output = Self;
+
+    fn div(self, rhs: i64) -> Self {
+        Self {
+            milliseconds: self.milliseconds / rhs,
+        }
+    }
+}
+
+/// Divide a `ShortInterval` by a numeric value, producing a `ShortInterval`.
+impl Div<F64> for ShortInterval {
+    type Output = Self;
+
+    fn div(self, rhs: F64) -> Self {
+        Self {
+            milliseconds: (F64::from(self.milliseconds as f64) / rhs).into_inner() as i64,
+        }
+    }
+}
+
+/// Divide a `ShortInterval` by a numeric value, producing a `ShortInterval`.
+impl Div<Decimal> for ShortInterval {
+    type Output = Self;
+
+    fn div(self, rhs: Decimal) -> Self {
+        Self {
+            milliseconds: (Decimal::from(self.milliseconds) / rhs)
+                .try_into()
+                .expect("overflow in short interval division"),
         }
     }
 }
@@ -230,17 +288,72 @@ impl LongInterval {
     }
 }
 
-impl<T> Mul<T> for LongInterval
-where
-    T: PrimInt,
-    i32: Mul<T, Output = i32>,
-{
+/// Multiply a `LongInterval` by an integer producing a `LongInterval`
+impl Mul<i32> for LongInterval {
     type Output = Self;
 
-    /// Multiply a long interval by an integer.
-    fn mul(self, rhs: T) -> Self {
+    fn mul(self, rhs: i32) -> Self {
         Self {
             months: self.months * rhs,
+        }
+    }
+}
+
+/// Multiply a `LongInterval` by a numeric value producing a `LongInterval`
+impl Mul<F64> for LongInterval {
+    type Output = Self;
+
+    fn mul(self, rhs: F64) -> Self {
+        Self {
+            months: (F64::from(self.months as f64) * rhs).into_inner() as i32,
+        }
+    }
+}
+
+/// Multiply a `LongInterval` by a numeric value producing a `LongInterval`
+impl Mul<Decimal> for LongInterval {
+    type Output = Self;
+
+    fn mul(self, rhs: Decimal) -> Self {
+        Self {
+            months: (Decimal::from(self.months) * rhs)
+                .try_into()
+                .expect("overflow in long interval multiplication"),
+        }
+    }
+}
+
+/// Divide a `LongInterval` by an integer producing a `LongInterval`
+impl Div<i32> for LongInterval {
+    type Output = Self;
+
+    fn div(self, rhs: i32) -> Self {
+        Self {
+            months: self.months / rhs,
+        }
+    }
+}
+
+/// Divide a `LongInterval` by a numeric value producing a `LongInterval`
+impl Div<F64> for LongInterval {
+    type Output = Self;
+
+    fn div(self, rhs: F64) -> Self {
+        Self {
+            months: (F64::from(self.months as f64) / rhs).into_inner() as i32,
+        }
+    }
+}
+
+/// Divide a `LongInterval` by a numeric value producing a `LongInterval`
+impl Div<Decimal> for LongInterval {
+    type Output = Self;
+
+    fn div(self, rhs: Decimal) -> Self {
+        Self {
+            months: (Decimal::from(self.months) / rhs)
+                .try_into()
+                .expect("overflow in long interval division"),
         }
     }
 }
@@ -299,3 +412,115 @@ some_operator!(eq, LongInterval, LongInterval, bool);
 some_operator!(neq, LongInterval, LongInterval, bool);
 some_operator!(gte, LongInterval, LongInterval, bool);
 some_operator!(lte, LongInterval, LongInterval, bool);
+
+#[doc(hidden)]
+pub fn times_ShortInterval_i64(left: ShortInterval, right: i64) -> ShortInterval {
+    left * right
+}
+
+some_polymorphic_function2!(times, ShortInterval, ShortInterval, i64, i64, ShortInterval);
+
+#[doc(hidden)]
+pub fn times_LongInterval_i32(left: LongInterval, right: i32) -> LongInterval {
+    left * right
+}
+
+some_polymorphic_function2!(times, LongInterval, LongInterval, i32, i32, LongInterval);
+
+#[doc(hidden)]
+pub fn times_ShortInterval_d(left: ShortInterval, right: F64) -> ShortInterval {
+    left * right
+}
+
+some_polymorphic_function2!(times, ShortInterval, ShortInterval, d, F64, ShortInterval);
+
+#[doc(hidden)]
+pub fn times_LongInterval_d(left: LongInterval, right: F64) -> LongInterval {
+    left * right
+}
+
+some_polymorphic_function2!(times, LongInterval, LongInterval, d, F64, LongInterval);
+
+#[doc(hidden)]
+pub fn times_ShortInterval_decimal(left: ShortInterval, right: Decimal) -> ShortInterval {
+    left * right
+}
+
+some_polymorphic_function2!(
+    times,
+    ShortInterval,
+    ShortInterval,
+    decimal,
+    Decimal,
+    ShortInterval
+);
+
+#[doc(hidden)]
+pub fn times_LongInterval_decimal(left: LongInterval, right: Decimal) -> LongInterval {
+    left * right
+}
+
+some_polymorphic_function2!(
+    times,
+    LongInterval,
+    LongInterval,
+    decimal,
+    Decimal,
+    LongInterval
+);
+
+#[doc(hidden)]
+pub fn div_ShortInterval_i64(left: ShortInterval, right: i64) -> ShortInterval {
+    left / right
+}
+
+some_polymorphic_function2!(div, ShortInterval, ShortInterval, i64, i64, ShortInterval);
+
+#[doc(hidden)]
+pub fn div_LongInterval_i32(left: LongInterval, right: i32) -> LongInterval {
+    left / right
+}
+
+some_polymorphic_function2!(div, LongInterval, LongInterval, i32, i32, LongInterval);
+
+#[doc(hidden)]
+pub fn div_ShortInterval_d(left: ShortInterval, right: F64) -> ShortInterval {
+    left / right
+}
+
+some_polymorphic_function2!(div, ShortInterval, ShortInterval, d, F64, ShortInterval);
+
+#[doc(hidden)]
+pub fn div_LongInterval_d(left: LongInterval, right: F64) -> LongInterval {
+    left / right
+}
+
+some_polymorphic_function2!(div, LongInterval, LongInterval, d, F64, LongInterval);
+
+#[doc(hidden)]
+pub fn div_ShortInterval_decimal(left: ShortInterval, right: Decimal) -> ShortInterval {
+    left / right
+}
+
+some_polymorphic_function2!(
+    div,
+    ShortInterval,
+    ShortInterval,
+    decimal,
+    Decimal,
+    ShortInterval
+);
+
+#[doc(hidden)]
+pub fn div_LongInterval_decimal(left: LongInterval, right: Decimal) -> LongInterval {
+    left / right
+}
+
+some_polymorphic_function2!(
+    div,
+    LongInterval,
+    LongInterval,
+    decimal,
+    Decimal,
+    LongInterval
+);

--- a/crates/sqllib/src/lib.rs
+++ b/crates/sqllib/src/lib.rs
@@ -1202,48 +1202,6 @@ where
 }
 
 #[doc(hidden)]
-pub fn times_ShortInterval_i64(left: ShortInterval, right: i64) -> ShortInterval {
-    left * right
-}
-
-some_polymorphic_function2!(times, ShortInterval, ShortInterval, i64, i64, ShortInterval);
-
-#[doc(hidden)]
-pub fn times_i64_ShortInterval(left: i64, right: ShortInterval) -> ShortInterval {
-    right * left
-}
-
-some_polymorphic_function2!(times, i64, i64, ShortInterval, ShortInterval, ShortInterval);
-
-#[doc(hidden)]
-pub fn times_ShortInterval_i32(left: ShortInterval, right: i32) -> ShortInterval {
-    left * (right as i64)
-}
-
-some_polymorphic_function2!(times, ShortInterval, ShortInterval, i32, i32, ShortInterval);
-
-#[doc(hidden)]
-pub fn times_i32_ShortInterval(left: i32, right: ShortInterval) -> ShortInterval {
-    right * (left as i64)
-}
-
-some_polymorphic_function2!(times, i32, i32, ShortInterval, ShortInterval, ShortInterval);
-
-#[doc(hidden)]
-pub fn times_i32_LongInterval(left: i32, right: LongInterval) -> LongInterval {
-    right * left
-}
-
-some_polymorphic_function2!(times, i32, i32, LongInterval, LongInterval, LongInterval);
-
-#[doc(hidden)]
-pub fn times_LongInterval_i32(left: LongInterval, right: i32) -> LongInterval {
-    left * right
-}
-
-some_polymorphic_function2!(times, LongInterval, LongInterval, i32, i32, LongInterval);
-
-#[doc(hidden)]
 #[inline(always)]
 pub fn truncate_d(left: F64, right: i32) -> F64 {
     let mut left = left.into_inner() * 10.0_f64.pow(right);

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/RustSqlRuntimeLibrary.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/RustSqlRuntimeLibrary.java
@@ -82,9 +82,10 @@ public class RustSqlRuntimeLibrary {
         this.arithmeticFunctions.put(DBSPOpcode.CONTROLLED_FILTER_GTE.toString(),
                 DBSPOpcode.CONTROLLED_FILTER_GTE);
 
-        this.dateFunctions.put("plus", DBSPOpcode.ADD);
-        this.dateFunctions.put("minus", DBSPOpcode.SUB);
-        this.dateFunctions.put("times", DBSPOpcode.MUL);
+        this.dateFunctions.put("plus", DBSPOpcode.TS_ADD);
+        this.dateFunctions.put("minus", DBSPOpcode.TS_SUB);
+        this.dateFunctions.put("times", DBSPOpcode.INTERVAL_MUL);
+        this.dateFunctions.put("div", DBSPOpcode.INTERVAL_DIV);
         this.dateFunctions.put(DBSPOpcode.CONTROLLED_FILTER_GTE.toString(), DBSPOpcode.CONTROLLED_FILTER_GTE);
 
         this.stringFunctions.put("concat", DBSPOpcode.CONCAT);
@@ -165,12 +166,10 @@ public class RustSqlRuntimeLibrary {
             map = this.universalFunctions;
         } else if (ltype.as(DBSPTypeBool.class) != null) {
             map = this.booleanFunctions;
-        } else if (ltype.is(IsDateType.class)) {
+        } else if (ltype.is(IsTimeRelatedType.class)) {
             map = this.dateFunctions;
-            if (opcode == DBSPOpcode.SUB || opcode == DBSPOpcode.ADD) {
-                if (ltype.is(DBSPTypeTimestamp.class) ||
-                        ltype.is(DBSPTypeDate.class) ||
-                        ltype.is(DBSPTypeTime.class)) {
+            if (opcode == DBSPOpcode.TS_SUB || opcode == DBSPOpcode.TS_ADD) {
+                if (ltype.is(IsTimeRelatedType.class)) {
                     assert rtype != null;
                     suffixReturn = "_" + expectedReturnType.baseTypeWithSuffix();
                     if (rtype.is(IsNumericType.class))
@@ -185,8 +184,8 @@ public class RustSqlRuntimeLibrary {
         } else if (ltype.is(DBSPTypeBinary.class) || ltype.is(DBSPTypeVariant.class)) {
             map = this.otherFunctions;
         }
-        if (rtype != null && rtype.is(IsDateType.class)) {
-            if (opcode.equals(DBSPOpcode.MUL)) {
+        if (rtype != null && rtype.is(IsIntervalType.class)) {
+            if (opcode == DBSPOpcode.INTERVAL_MUL || opcode == DBSPOpcode.INTERVAL_DIV) {
                 // e.g., 10 * INTERVAL
                 map = this.dateFunctions;
             }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
@@ -802,7 +802,7 @@ public class ToRustInnerVisitor extends InnerVisitor {
 
         functionName = "cast_to_" + destType.baseTypeWithSuffix() +
                 "_" + sourceType.baseTypeWithSuffix();
-        this.builder.append(functionName).append("(").increase();
+        this.builder.append(functionName).append("(");
         expression.source.accept(this);
         DBSPTypeDecimal dec = destType.as(DBSPTypeDecimal.class);
         if (dec != null) {
@@ -820,7 +820,7 @@ public class ToRustInnerVisitor extends InnerVisitor {
                     .append(", ")
                     .append(Boolean.toString(str.fixed));
         }
-        this.builder.decrease().append(")");
+        this.builder.append(")");
         return VisitDecision.STOP;
     }
 
@@ -937,11 +937,11 @@ public class ToRustInnerVisitor extends InnerVisitor {
                         expression.getType(),
                         expression.left.getType(),
                         expression.right.getType());
-                this.builder.append(function).append("(").increase();
+                this.builder.append(function).append("(");
                 expression.left.accept(this);
-                this.builder.append(",").newline();
+                this.builder.append(", ");
                 expression.right.accept(this);
-                this.builder.decrease().append(")");
+                this.builder.append(")");
                 break;
             }
         }
@@ -1401,7 +1401,7 @@ public class ToRustInnerVisitor extends InnerVisitor {
             return this.doNullExpression(expression);
         if (expression.getType().mayBeNull)
             this.builder.append("Some(");
-        boolean newlines = this.compact && expression.fields.length > 2;
+        boolean newlines = !this.compact && expression.fields.length > 2;
         this.builder.append(DBSPTypeCode.TUPLE.rustName)
                 .append(expression.size())
                 .append("::new(");

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
@@ -86,9 +86,12 @@ import org.dbsp.sqlCompiler.ir.expression.literal.DBSPU32Literal;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPVecLiteral;
 import org.dbsp.sqlCompiler.ir.path.DBSPPath;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
+import org.dbsp.sqlCompiler.ir.type.IsDateType;
+import org.dbsp.sqlCompiler.ir.type.IsIntervalType;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeAny;
 import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeRef;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeVariant;
+import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeIndexedZSet;
 import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeMap;
 import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeResult;
 import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeStruct;
@@ -96,7 +99,7 @@ import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeTuple;
 import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeTupleBase;
 import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeUser;
 import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeVec;
-import org.dbsp.sqlCompiler.ir.type.IsDateType;
+import org.dbsp.sqlCompiler.ir.type.IsTimeRelatedType;
 import org.dbsp.sqlCompiler.ir.type.IsNumericType;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBinary;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBool;
@@ -357,8 +360,8 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
     public static boolean needCommonType(DBSPOpcode opcode, DBSPType result, DBSPType left, DBSPType right) {
         if (opcode == DBSPOpcode.CONCAT) return false;
         // Dates can be mixed with other types in a binary operation
-        if (left.is(IsDateType.class)) return false;
-        if (right.is(IsDateType.class)) return false;
+        if (left.is(IsTimeRelatedType.class)) return false;
+        if (right.is(IsTimeRelatedType.class)) return false;
         // Allow arithmetic on different DECIMAL types
         if (left.is(DBSPTypeDecimal.class) && right.is(DBSPTypeDecimal.class))
             return false;
@@ -401,6 +404,16 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
                 return new DBSPUnaryExpression(node, resultType, DBSPOpcode.INDICATOR, argument.borrow());
             }
         }
+    }
+
+    static DBSPOpcode timestampOperation(DBSPOpcode opcode) {
+        return switch (opcode) {
+            case ADD -> DBSPOpcode.TS_ADD;
+            case SUB -> DBSPOpcode.TS_SUB;
+            case MUL -> DBSPOpcode.INTERVAL_MUL;
+            case DIV -> DBSPOpcode.INTERVAL_DIV;
+            default -> throw new InternalCompilerError("Unexpected opcode " + opcode);
+        };
     }
 
     public static DBSPExpression makeBinaryExpression(
@@ -452,14 +465,57 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
                 // Result is always NULL - evaluate to the NULL literal directly
                 return DBSPLiteral.none(type);
             }
+            if (opcode == DBSPOpcode.MUL || opcode == DBSPOpcode.DIV) {
+                // Multiplication between an interval and a numeric value.
+                if (leftType.is(IsIntervalType.class) || rightType.is(IsIntervalType.class)) {
+                    opcode = timestampOperation(opcode);
+                    // swap operands so that the numeric operand is always right
+                    if (opcode == DBSPOpcode.INTERVAL_MUL || opcode == DBSPOpcode.INTERVAL_DIV) {
+                        if (rightType.is(IsIntervalType.class)) {
+                            if (leftType.is(IsIntervalType.class)) {
+                                throw new CompilationError("Operation " + opcode + " between intervals not supported", node);
+                            }
+                            DBSPExpression tmp = left;
+                            left = right;
+                            right = tmp;
+
+                            leftType = left.getType();
+                            rightType = right.getType();
+                        }
+
+                        assert rightType.is(IsNumericType.class);
+                        // Canonicalize the type on the right without information loss
+                        if (rightType.is(DBSPTypeInteger.class)) {
+                            if (leftType.is(DBSPTypeMillisInterval.class)) {
+                                right = right.cast(new DBSPTypeInteger(rightType.getNode(), 64, true, rightType.mayBeNull));
+                            } else {
+                                right = right.cast(new DBSPTypeInteger(rightType.getNode(), 32, true, rightType.mayBeNull));
+                            }
+                        } else if (rightType.is(DBSPTypeReal.class)) {
+                            right = right.cast(new DBSPTypeDouble(rightType.getNode(), rightType.mayBeNull));
+                        }
+                        rightType = right.getType();
+                    }
+                }
+            }
             if (opcode == DBSPOpcode.SUB || opcode == DBSPOpcode.ADD) {
-                if (leftType.is(DBSPTypeTimestamp.class) || leftType.is(DBSPTypeDate.class)) {
+                // Addition involving a date
+                if (leftType.is(IsTimeRelatedType.class) || rightType.is(IsTimeRelatedType.class)) {
                     if (rightType.is(IsNumericType.class))
                         throw new CompilationError("Cannot apply operation " + Utilities.singleQuote(opcode.toString()) +
                                 " to arguments of type " + leftType.asSqlString() + " and " + rightType.asSqlString(), node);
+                    opcode = timestampOperation(opcode);
+                    if (leftType.is(IsIntervalType.class) && !rightType.is(IsIntervalType.class)) {
+                        // Move the interval to the right when computing a date +/- interval
+                        DBSPExpression tmp = left;
+                        left = right;
+                        right = tmp;
+                        leftType = left.getType();
+                        rightType = right.getType();
+                    }
                 }
             }
-            if (leftType.is(IsDateType.class) || rightType.is(IsDateType.class))
+            if (leftType.is(IsTimeRelatedType.class) || rightType.is(IsTimeRelatedType.class))
                 expressionResultType = typeWithNull;
             else if (leftType.is(DBSPTypeString.class) && rightType.is(DBSPTypeString.class))
                 expressionResultType = typeWithNull;

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/Simplify.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/Simplify.java
@@ -509,7 +509,7 @@ public class Simplify extends InnerRewriteVisitor {
                         }
                     }
                 }
-            } else if (expression.operation.equals(DBSPOpcode.ADD)) {
+            } else if (expression.operation == DBSPOpcode.ADD || expression.operation == DBSPOpcode.TS_ADD) {
                 if (left.is(DBSPLiteral.class)) {
                     DBSPLiteral leftLit = left.to(DBSPLiteral.class);
                     IHasZero iLeftType = leftType.as(IHasZero.class);
@@ -532,7 +532,7 @@ public class Simplify extends InnerRewriteVisitor {
                         }
                     }
                 }
-            } else if (expression.operation.equals(DBSPOpcode.MUL)) {
+            } else if (expression.operation == DBSPOpcode.MUL || expression.operation == DBSPOpcode.INTERVAL_MUL) {
                 if (left.is(DBSPLiteral.class) && leftType.is(IsNumericType.class)) {
                     DBSPLiteral leftLit = left.to(DBSPLiteral.class);
                     IsNumericType iLeftType = leftType.to(IsNumericType.class);
@@ -565,7 +565,7 @@ public class Simplify extends InnerRewriteVisitor {
                                 .to(DBSPExpression.class);
                     }
                 }
-            } else if (expression.operation.equals(DBSPOpcode.DIV)) {
+            } else if (expression.operation == DBSPOpcode.DIV || expression.operation == DBSPOpcode.INTERVAL_DIV) {
                 if (right.is(DBSPLiteral.class)) {
                     DBSPLiteral rightLit = right.to(DBSPLiteral.class);
                     IsNumericType iRightType = rightType.to(IsNumericType.class);
@@ -576,7 +576,7 @@ public class Simplify extends InnerRewriteVisitor {
                                 " Division by constant zero value.");
                     }
                 }
-            } else if (expression.operation.equals(DBSPOpcode.MOD)) {
+            } else if (expression.operation == DBSPOpcode.MOD) {
                 if (right.is(DBSPLiteral.class)) {
                     DBSPLiteral rightLit = right.to(DBSPLiteral.class);
                     IsNumericType iRightType = rightType.to(IsNumericType.class);

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/monotone/MonotoneTransferFunctions.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/monotone/MonotoneTransferFunctions.java
@@ -531,7 +531,8 @@ public class MonotoneTransferFunctions extends TranslateVisitor<MonotoneExpressi
 
         // Assume type is not monotone
         IMaybeMonotoneType resultType = new NonMonotoneType(expression.type);
-        if (expression.operation == DBSPOpcode.ADD && lm && rm) {
+        if ((expression.operation == DBSPOpcode.ADD || expression.operation == DBSPOpcode.TS_ADD)
+                && lm && rm) {
             // The addition of two monotone expressions is monotone
             resultType = new MonotoneType(expression.type);
             reduced = expression.replaceSources(
@@ -560,7 +561,9 @@ public class MonotoneTransferFunctions extends TranslateVisitor<MonotoneExpressi
                     right.getReducedExpression());
         }
         // Some expressions are monotone if some of their operands are constant
-        if (left.mayBeMonotone() && expression.operation == DBSPOpcode.SUB) {
+        if (left.mayBeMonotone() &&
+                (expression.operation == DBSPOpcode.SUB ||
+                        expression.operation == DBSPOpcode.TS_SUB)) {
             // Subtracting a constant from a monotone expression produces a monotone result
             if (this.constantExpressions.contains(expression.right)) {
                 resultType = left.copyMonotonicity(expression.type);
@@ -569,7 +572,10 @@ public class MonotoneTransferFunctions extends TranslateVisitor<MonotoneExpressi
             }
         }
         if (left.mayBeMonotone() &&
-                (expression.operation == DBSPOpcode.DIV || expression.operation == DBSPOpcode.MUL)) {
+                (expression.operation == DBSPOpcode.DIV ||
+                        expression.operation == DBSPOpcode.MUL ||
+                        expression.operation == DBSPOpcode.INTERVAL_MUL ||
+                        expression.operation == DBSPOpcode.INTERVAL_DIV)) {
             // Multiplying or dividing a monotone expression by
             // a positive constant produces a monotone result
             // TODO: multiplication is commutative.

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/monotonicity/Monotonicity.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/monotonicity/Monotonicity.java
@@ -950,6 +950,8 @@ public class Monotonicity extends CircuitVisitor {
 
         static DBSPOpcode inverse(DBSPOpcode opcode) {
             return switch (opcode) {
+                case TS_ADD -> DBSPOpcode.TS_SUB;
+                case TS_SUB -> DBSPOpcode.TS_ADD;
                 case ADD -> DBSPOpcode.SUB;
                 case SUB -> DBSPOpcode.ADD;
                 case LT -> DBSPOpcode.GTE;
@@ -968,7 +970,8 @@ public class Monotonicity extends CircuitVisitor {
             DBSPBinaryExpression binary = larger.as(DBSPBinaryExpression.class);
             if (binary == null)
                 return false;
-            if (binary.operation != DBSPOpcode.ADD && binary.operation != DBSPOpcode.SUB)
+            if (binary.operation != DBSPOpcode.ADD && binary.operation != DBSPOpcode.SUB &&
+                binary.operation != DBSPOpcode.TS_ADD && binary.operation != DBSPOpcode.TS_SUB)
                 return false;
             DBSPOpcode inverse = inverse(binary.operation);
             int column = isColumn(binary.left, param);
@@ -982,7 +985,7 @@ public class Monotonicity extends CircuitVisitor {
                 return true;
             }
 
-            if (binary.operation != DBSPOpcode.ADD)
+            if ((binary.operation == DBSPOpcode.SUB) || (binary.operation == DBSPOpcode.TS_SUB))
                 return false;
 
             // constant + col

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPOpcode.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPOpcode.java
@@ -50,6 +50,13 @@ public enum DBSPOpcode {
     // Shift amount may be negative
     SHIFT_LEFT("shift_left", false),
 
+    // Timestamp-based operations
+    TS_ADD("+", false),
+    TS_SUB("-", false),
+    // Interval-based operations
+    INTERVAL_MUL("*", false),
+    INTERVAL_DIV("/", false),
+
     // Aggregate operations.  These operations
     // handle NULL values differently from standard
     // arithmetic operations, following SQL semantics.

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/IsDateType.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/IsDateType.java
@@ -23,5 +23,10 @@
 
 package org.dbsp.sqlCompiler.ir.type;
 
-/** interface implemented by types that look like dates. */
+import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeDate;
+import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeTime;
+import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeTimestamp;
+
+/** interface implemented by {@link DBSPTypeTimestamp}, {@link DBSPTypeDate}, 
+ * {@link DBSPTypeTime}. */
 public interface IsDateType extends IsBoundedType {}

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/IsIntervalType.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/IsIntervalType.java
@@ -1,0 +1,4 @@
+package org.dbsp.sqlCompiler.ir.type;
+
+/** interface implemented by Interval types. */
+public interface IsIntervalType extends IsBoundedType {}

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/IsTimeRelatedType.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/IsTimeRelatedType.java
@@ -1,0 +1,4 @@
+package org.dbsp.sqlCompiler.ir.type;
+
+/** interface implemented by {@link IsDateType} and interval types. */
+public interface IsTimeRelatedType extends IsDateType {}

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeDate.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeDate.java
@@ -31,12 +31,13 @@ import org.dbsp.sqlCompiler.ir.expression.DBSPExpression;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPDateLiteral;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
 import org.dbsp.sqlCompiler.ir.type.IsDateType;
+import org.dbsp.sqlCompiler.ir.type.IsTimeRelatedType;
 
 import java.util.Objects;
 
 import static org.dbsp.sqlCompiler.ir.type.DBSPTypeCode.DATE;
 
-public class DBSPTypeDate extends DBSPTypeBaseType implements IsDateType {
+public class DBSPTypeDate extends DBSPTypeBaseType implements IsDateType, IsTimeRelatedType {
     public DBSPTypeDate(CalciteObject node, boolean mayBeNull) {
         super(node, DATE, mayBeNull);
     }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeMillisInterval.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeMillisInterval.java
@@ -31,7 +31,8 @@ import org.dbsp.sqlCompiler.ir.expression.literal.DBSPIntervalMillisLiteral;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPLiteral;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
 import org.dbsp.sqlCompiler.ir.type.IHasZero;
-import org.dbsp.sqlCompiler.ir.type.IsDateType;
+import org.dbsp.sqlCompiler.ir.type.IsIntervalType;
+import org.dbsp.sqlCompiler.ir.type.IsTimeRelatedType;
 
 import java.util.Objects;
 
@@ -41,7 +42,7 @@ import static org.dbsp.sqlCompiler.ir.type.DBSPTypeCode.INTERVAL_SHORT;
  * Always stores the interval value in milliseconds. */
 public class DBSPTypeMillisInterval
         extends DBSPTypeBaseType
-        implements IsDateType, IHasZero {
+        implements IsTimeRelatedType, IHasZero, IsIntervalType {
     public DBSPTypeMillisInterval(CalciteObject node, boolean mayBeNull) {
         super(node, INTERVAL_SHORT, mayBeNull);
     }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeMonthsInterval.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeMonthsInterval.java
@@ -32,7 +32,8 @@ import org.dbsp.sqlCompiler.ir.expression.literal.DBSPLiteral;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
 import org.dbsp.sqlCompiler.ir.type.DBSPTypeCode;
 import org.dbsp.sqlCompiler.ir.type.IHasZero;
-import org.dbsp.sqlCompiler.ir.type.IsDateType;
+import org.dbsp.sqlCompiler.ir.type.IsIntervalType;
+import org.dbsp.sqlCompiler.ir.type.IsTimeRelatedType;
 
 import java.util.Objects;
 
@@ -40,7 +41,7 @@ import java.util.Objects;
  * Always stores the interval value in days. */
 public class DBSPTypeMonthsInterval
         extends DBSPTypeBaseType
-        implements IsDateType, IHasZero {
+        implements IsTimeRelatedType, IHasZero, IsIntervalType {
     public DBSPTypeMonthsInterval(CalciteObject node, boolean mayBeNull) {
         super(node, DBSPTypeCode.INTERVAL_LONG, mayBeNull);
     }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeTime.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeTime.java
@@ -31,12 +31,13 @@ import org.dbsp.sqlCompiler.ir.expression.DBSPExpression;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPTimeLiteral;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
 import org.dbsp.sqlCompiler.ir.type.IsDateType;
+import org.dbsp.sqlCompiler.ir.type.IsTimeRelatedType;
 
 import java.util.Objects;
 
 import static org.dbsp.sqlCompiler.ir.type.DBSPTypeCode.TIME;
 
-public class DBSPTypeTime extends DBSPTypeBaseType implements IsDateType {
+public class DBSPTypeTime extends DBSPTypeBaseType implements IsDateType, IsTimeRelatedType {
     public DBSPTypeTime(CalciteObject node, boolean mayBeNull) {
         super(node, TIME, mayBeNull);
     }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeTimestamp.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/type/primitive/DBSPTypeTimestamp.java
@@ -30,13 +30,14 @@ import org.dbsp.sqlCompiler.ir.expression.DBSPExpression;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPTimestampLiteral;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
 import org.dbsp.sqlCompiler.ir.type.IsDateType;
+import org.dbsp.sqlCompiler.ir.type.IsTimeRelatedType;
 
 import java.util.Objects;
 
 import static org.dbsp.sqlCompiler.ir.type.DBSPTypeCode.TIMESTAMP;
 
 public class DBSPTypeTimestamp extends DBSPTypeBaseType
-        implements IsDateType {
+        implements IsDateType, IsTimeRelatedType {
     public DBSPTypeTimestamp(CalciteObject node, boolean mayBeNull) {
         super(node, TIMESTAMP, mayBeNull);
     }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/mysql/TimestampdiffTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/mysql/TimestampdiffTests.java
@@ -4,6 +4,15 @@ import org.dbsp.sqlCompiler.compiler.sql.tools.SqlIoTest;
 import org.junit.Test;
 
 public class TimestampdiffTests extends SqlIoTest {
+    @Test
+    public void issue3021() {
+        this.compileRustTestCase("""
+                CREATE TABLE time_tbl(c1 TIME);
+                CREATE MATERIALIZED VIEW v AS SELECT
+                c1 - INTERVAL '10' MINUTE AS f_c1
+                FROM time_tbl;""");
+    }
+
     // Test data obtained from
     // https://github.com/mysql/mysql-server/blob/mysql-test/r/func_time.result#L715
     @Test

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/quidem/MiscTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/quidem/MiscTests.java
@@ -65,7 +65,8 @@ public class MiscTests extends ScottBaseTests {
                 (1 row)""");
     }
 
-    @Test @Ignore("Requires MULTISET")
+    @Test
+    @Ignore("Requires MULTISET")
     public void testRowCoalesce() {
         this.qs("""
                 -- [CALCITE-877] Allow ROW as argument to COLLECT
@@ -81,5 +82,87 @@ public class MiscTests extends ScottBaseTests {
                 |     30 | [{7499, 30}, {7521, 30}, {7654, 30}, {7698, 30}, {7844, 30}, {7900, 30}] |
                 +--------+--------------------------------------------------------------------------+
                 (3 rows)""");
+    }
+
+    @Test
+    public void intervalTests() {
+        // Added tests with decimal and FP
+        this.qs("""
+                -- [CALCITE-922] Value of INTERVAL literal
+                select deptno * interval '2' day as d2,
+                 deptno * interval -'3' hour as h3,
+                 deptno * interval -'-4' hour as h4,
+                 deptno * interval -'4:30' hour to minute as h4_5,
+                 deptno * interval -'-1-3' year to month as y1_25,
+                 CAST(deptno AS DECIMAL(6, 2)) / 10 * interval 1 minutes as m,
+                 CAST(deptno AS REAL) / 15 * interval 1 day as d
+                from dept;
+                +---------+------------+-----------+------------+--------------------+--------+-----------------+
+                | D2      | H3         | H4        | H4_5       | Y1_25              | M      | D               |
+                +---------+------------+-----------+------------+--------------------+--------+-----------------+
+                | 20 days |  -30 hours |  40 hours |  -45 hours | 12 years 06 months | 1 mins |  57600.001 secs |
+                | 40 days |  -60 hours |  80 hours |  -90 hours | 25 years 00 months | 2 mins | 115200.003 secs |
+                | 60 days |  -90 hours | 120 hours | -135 hours | 37 years 06 months | 3 mins | 172800 secs     |
+                | 80 days | -120 hours | 160 hours | -180 hours | 50 years 00 months | 4 mins | 230400.006 secs |
+                +---------+------------+-----------+------------+--------------------+--------+-----------------+
+                (4 rows)
+                
+                -- [CALCITE-4091] Interval expressions
+                select empno, mgr, date '1970-01-01' + interval empno day as d,
+                  timestamp '1970-01-01 00:00:00' + interval (mgr / 100) minute as ts
+                from emp
+                order by empno;
+                +-------+------+------------+---------------------+
+                | EMPNO | MGR  | D          | TS                  |
+                +-------+------+------------+---------------------+
+                |  7369 | 7902 | 1990-03-06 | 1970-01-01 01:19:00 |
+                |  7499 | 7698 | 1990-07-14 | 1970-01-01 01:16:00 |
+                |  7521 | 7698 | 1990-08-05 | 1970-01-01 01:16:00 |
+                |  7566 | 7839 | 1990-09-19 | 1970-01-01 01:18:00 |
+                |  7654 | 7698 | 1990-12-16 | 1970-01-01 01:16:00 |
+                |  7698 | 7839 | 1991-01-29 | 1970-01-01 01:18:00 |
+                |  7782 | 7839 | 1991-04-23 | 1970-01-01 01:18:00 |
+                |  7788 | 7566 | 1991-04-29 | 1970-01-01 01:15:00 |
+                |  7839 |      | 1991-06-19 |                     |
+                |  7844 | 7698 | 1991-06-24 | 1970-01-01 01:16:00 |
+                |  7876 | 7788 | 1991-07-26 | 1970-01-01 01:17:00 |
+                |  7900 | 7698 | 1991-08-19 | 1970-01-01 01:16:00 |
+                |  7902 | 7566 | 1991-08-21 | 1970-01-01 01:15:00 |
+                |  7934 | 7782 | 1991-09-22 | 1970-01-01 01:17:00 |
+                +-------+------+------------+---------------------+
+                (14 rows)
+                
+                -- [CALCITE-6581] INTERVAL with WEEK and QUARTER
+                select timestamp '1970-01-01' + interval '2' week as w,
+                  timestamp '1970-01-01 00:00:00' + interval '2' quarter as q;
+                +---------------------+---------------------+
+                | W                   | Q                   |
+                +---------------------+---------------------+
+                | 1970-01-15 00:00:00 | 1970-07-01 00:00:00 |
+                +---------------------+---------------------+
+                (1 row)
+                """);
+    }
+
+    @Test
+    public void intervalDivision() {
+        // Tested on Postgres, the long interval results differ, since postgres computes on days
+        this.qs("""
+                select interval '2' day / deptno as d2,
+                 interval -'3' hour / deptno as h3,
+                 interval -'-4' hour / deptno as h4,
+                 interval -'4:30' hour to minute / deptno as h4_5,
+                 interval -'-1-3' year to month / deptno as y1_25,
+                 interval 1 minutes / CAST(deptno AS DECIMAL(6, 2)) as m
+                from dept;
+                +-----------------+-----------------+---------+-------------------+----------+----------+
+                | D2              | H3              | H4      | H4_5              | Y1_25    | M        |
+                +-----------------+-----------------+---------+-------------------+----------+----------+
+                | 4 hours 48 mins |        -18 mins | 24 mins |          -27 mins | 1 months |   6 secs |
+                | 2 hours 24 mins |         -9 mins | 12 mins |  -13 mins 30 secs | 0 months |   3 secs |
+                | 1 hour 36 mins  |         -6 mins |  8 mins |           -9 mins | 0 months |   2 secs |
+                | 1 hour 12 mins  | -4 mins 30 secs |  6 mins |   -6 mins 45 secs | 0 months | 1.5 secs |
+                +-----------------+-----------------+---------+-------------------+----------+----------+
+                (4 rows)""");
     }
 }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/tools/TableParser.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/tools/TableParser.java
@@ -68,7 +68,7 @@ public class TableParser {
     static final SimpleDateFormat DATE_OUTPUT_FORMAT = new SimpleDateFormat("yyyy-MM-dd");
     static final Pattern YEAR = Pattern.compile("^(\\d+) years?(.*)");
     static final Pattern MONTHS = Pattern.compile("^\\s*(\\d+) months?(.*)");
-    static final Pattern MINUS = Pattern.compile("^-(.*)");
+    static final Pattern MINUS = Pattern.compile("^-\\s*(.*)");
     static final Pattern DAYS = Pattern.compile("^(\\d+) days?(.*)");
     static final Pattern HOURS = Pattern.compile("^\\s*(\\d+) hours?(.*)");
     static final Pattern MINUTES = Pattern.compile("\\s*(\\d+) mins?(.*)");
@@ -194,22 +194,26 @@ public class TableParser {
             } else {
                 m = HOURS.matcher(interval);
                 if (m.matches()) {
-                    long h = Integer.parseInt(m.group(1));
+                    long h = Long.parseLong(m.group(1));
                     result += h * 3600_000;
                     interval = m.group(2);
                 }
 
                 m = MINUTES.matcher(interval);
                 if (m.matches()) {
-                    long mm = Integer.parseInt(m.group(1));
+                    long mm = Long.parseLong(m.group(1));
                     result += mm * 60_000;
                     interval = m.group(2);
                 }
 
                 m = SECONDS.matcher(interval);
                 if (m.matches()) {
-                    long s = Integer.parseInt(m.group(1));
+                    long s = Long.parseLong(m.group(1));
                     result += s * 1000;
+                    if (m.group(3) != null) {
+                        String msec = m.group(3) + "000";
+                        result += Long.parseLong(msec.substring(0, 3));
+                    }
                     interval = m.group(4);
                 }
             }

